x-pack/legacy/plugins/actions/server/builtin_action_types/webhook.ts

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import axios, { AxiosError, AxiosResponse } from 'axios';
+import { i18n } from '@kbn/i18n';
+import { schema, TypeOf } from '@kbn/config-schema';
+import { nullableType } from './lib/nullable';
+import { portSchema } from './lib/schemas';
+import { isOk, promiseResult, Result } from './lib/result_type';
+import { ActionType, ActionTypeExecutorOptions, ActionTypeExecutorResult } from '../types';
+
+// config definition
+enum WebhookMethods {
+  POST = 'post',
+  PUT = 'put',
+}
+enum WebhookSchemes {
+  HTTP = 'http',
+  HTTPS = 'https',
+}
+
+export type ActionTypeConfigType = TypeOf<typeof ConfigSchema>;
+
+const HeadersSchema = schema.recordOf(schema.string(), schema.string());
+
+export const CompositeUrlSchema = schema.object({
+  host: schema.string(),
+  port: portSchema(),
+  path: nullableType(schema.string()),
+  scheme: schema.oneOf(
+    [schema.literal(WebhookSchemes.HTTP), schema.literal(WebhookSchemes.HTTPS)],
+    {
+      defaultValue: WebhookSchemes.HTTP,
+    }
+  ),
+});
+
+const ConfigSchema = schema.object({
+  url: schema.oneOf([schema.string(), CompositeUrlSchema]),
+  method: schema.oneOf([schema.literal(WebhookMethods.POST), schema.literal(WebhookMethods.PUT)], {
+    defaultValue: WebhookMethods.POST,
+  }),
+  headers: nullableType(HeadersSchema),
+});
+
+// secrets definition
+export type ActionTypeSecretsType = TypeOf<typeof SecretsSchema>;
+const SecretsSchema = schema.object({
```

Comment:
Yup, the authtoken (which is already done locally, haven't pushed yet as still having trouble integration testing it) is an extension to this type using a "oneOf" relationship.

I didn't follow the validation point - is there a reason why using the typesystem to enforce both values being specified isn't good enough? Ideally I'd like to use the typesystem to enforce logic rather than imperative code as  it catches the invalidity earlier and is generally easier to maintain.

The only thing I can think of that might make this an issue is what you mentioned about the savedObjects semantics... which I don't know enough about yet, so if I'm missing something here, please do say. :)

Reply:
That might be true at runtime, but doesn't doing it this way means that at compile time were type safe too?
The benefit would be that when we interact with our config object it's type safe and if you wish to trigger an action programmatically from within Kibana, not via the API, you still have types.

Reply:
For me, it comes down to whether I want to wrestle with the type system vs doing some explicit checks.  One of the downsides of using config-schema is that there may be some unknown unknowns.

It's super-nice to be able to use a type from a schema, but I'm most interested in just getting the properties with the right types.  Extra validation beyond that doesn't really need to be part of that contract.  Here's the email config validation: https://github.com/elastic/kibana/blob/4dc7b3603ccfa0dc3a042ef718a245601aeb5472/x-pack/legacy/plugins/actions/server/builtin_action_types/email.ts#L34-L61

I'm happy with doing some extra validation like that.

OTOH, I'm also open to using union types for [{user,password}, {authToken}], to see what it looks like, once we get there.  That can be in a different PR, if so, create an issue or project card to track.

Reply:
Yeah, it might be a matter of taste.

In my book, the the main value in types is that it gives the developer immediate feedback so they know whether what they're doing is legal or not.
If the type doesn't let you create an invalid object, you know immediately.
And since type systems are (usually) harder to break than business logic, I prefer to use it to express the constraints over imperative logic that is more likely to be changed when I, or someone else, has forgotten the context of the constraints.

But, as I said, it's a matter of taste.
I think I've become so accustomed to typed code now that I definitely feel more confident in them, than in my own BL.

Reply:
My ideal is that the "JSON -> Typed" conversion happens right at the edge (as in, when the JSON input enters the server) and from there on out, throughout your pipeline, its using Typescript so that it's impossible to misuse the input.

But that's much harder to do when introducing types _into_ an existing system.